File: tests/r_parity/49_oprobit.py
"""StatsPAI ordered probit parity (Python side) -- Module 49."""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "49_oprobit"

def make_data(n=1000, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x = rng.normal(0, 1, n)
    ystar = 0.7 * x + rng.normal(0, 1, n)  # Gaussian errors -> probit link
    y = np.where(ystar < -0.5, 0, np.where(ystar < 0.5, 1, 2))
    return pd.DataFrame({"y": y, "x": x.astype(float)})

def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.oprobit(formula="y ~ x", data=df)
    rows = []
    for nm, lab in [("x", "x"), ("/cut1", "cut1"), ("/cut2", "cut2")]:
        if nm in res.params.index:
            rows.append(ParityRecord(MODULE, "py", f"beta_{lab}",
                estimate=float(res.params[nm]),
                se=float(res.std_errors[nm]),
                n=int(len(df))))
    write_results(MODULE, "py", rows, extra={"link": "probit", "engine": "statsmodels"})

if __name__ == "__main__":
    main()

File: tests/r_parity/49_oprobit.R
# StatsPAI ordered probit parity (R side) -- Module 49.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(MASS) })

MODULE <- "49_oprobit"
df <- read_csv_strict(MODULE)
df$y <- factor(df$y, ordered = TRUE)

fit <- MASS::polr(y ~ x, data = df, method = "probit", Hess = TRUE)
sm <- summary(fit)
co <- sm$coefficients

beta_x  <- co["x", "Value"];   se_x    <- co["x", "Std. Error"]
cut1    <- co["0|1", "Value"]; se_cut1 <- co["0|1", "Std. Error"]
cut2    <- co["1|2", "Value"]; se_cut2 <- co["1|2", "Std. Error"]

rows <- list(
  parity_row(MODULE, "beta_x",    estimate = beta_x,  se = se_x,    n = nrow(df)),
  parity_row(MODULE, "beta_cut1", estimate = cut1,    se = se_cut1, n = nrow(df)),
  parity_row(MODULE, "beta_cut2", estimate = cut2,    se = se_cut2, n = nrow(df))
)
write_results(MODULE, rows, extra = list(link = "probit", package = "MASS::polr"))
